raft: scaffolding for store liveness and leader leases

nvb opened the pull request:
We can use this to split up the prototyping effort. There's work to do below the new StoreLiveness interface, in Raft using StoreLiveness, and above Raft using the new Status.LeadSupportUntil field.

To collaborate on this, we can push new commits to this branch. Try to avoid force pushing to prevent skew.

nvb committed:
raft: scaffolding for store liveness and leader leases

We can use this to split up the prototyping effort. There's work to do
below the new StoreLiveness interface, in Raft using StoreLiveness, and
above Raft using the new Status.LeadSupportUntil field.

To collaborate on this, we can push new commits to this branch. Try to
avoid force pushing to prevent skew.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123789)
<!-- Reviewable:end -->

nvb committed:
kv: plumb replicaRLockedStoreLiveness wrapper into raft

This commit creates a new replicaRLockedStoreLiveness intermediary type
which translates replica IDs to store IDs using an rlocked replica's
RangeDescriptor and uses the resulting store ID to call into a real (and
as-of-yet unimplemented) StoreLiveness instance.

nvb committed:
kv: add Enabled method to StoreLiveness interface

nvb committed:
kv: integrate leader leases over raft leader support

arulajmani committed:
raft: add fortification handshake

This patch adds and implements two new message types -- MsgFortify and
MsgFortifyResp. Using these, the leader broadcasts a fortification
request to all its followers when a vote is won. It handles responses
by updating its leadSupport map.

Things still missing -
1. We're not evaluating if we're supported by a majority of followers,
and till when.
2. We're not persisting lead and leadEpoch to disk.
3. Re-fortification isn't handled.
4. Testing.

Release note: None

arulajmani committed:
raft: mock store liveness in datadriven tests

This patch mocks store liveness in datadriven tests and adds a few
directives to bump epochs and withdraw support. It then constructs
an "interesting" store liveness state and runs a new election. In doing
so, we ensure that MsgFortifyResp are populated correctly based on the
store liveness state.

Release note: None

arulajmani committed:
raft: enable periodic re-fortification

This patch build re-fortification. Now, on every tick, the leader tries
to re-fortify any peers that were never fortified or who have withdrawn
support from a fortified epoch. We only do this if the peer's store is
currently supporting the leader's; otherwise, we'll have to wait until
this happens.

Release note: None

nvb committed:
kv: propagate HLC timestamps across RaftTransport

This commit adds a ClockTimestamp to RaftMessageRequestBatch.

nvb committed:
raft: clean up fortify message construction

miraradeva committed:
storeliveness: add basic heartbeating functionality

This patch is based on Sumeer's prototype in
https://github.com/cockroachdb/cockroach/pull/122547. The main
differences are many simplifications due to the algorithm
changes between take 3 and take 4.

This patch adds only the basic heartbeating capability and structure of
the store liveness fabric. The algorithms logic will come in separate
commits.

miraradeva committed:
storeliveness: streaming transport and algorithm logic

Two main changes:
- The transport is not longer RPC, but streaming instead, with a
  per-node send queue for outgoing heartbeat requests, and separate
  per-store receive queues for incoming heartbeat requests, and
  responses.
- The algorithm logic is in a very basic state. Persistence is not
  implemented yet (there are TODOs). There is zero testing.

nvb committed:
storeliveness: rework transport and protocol

This makes the transport look more like etcd/raft.

nvb committed:
storeliveness: add persistence

nvb committed:
kv: respect MinExpiration

arulajmani committed:
raft: introduce SupportTracker

This patch introduces a new SupportTracker struct that tracks support
provided by followers to a leader. The leader can then use this tracked
support to calculate its QSE, which is used by higher layers.

Release note: None

nvb committed:
kv: fix merge skew, disable quiescence

arulajmani committed:
raft: ensure we campaign iff we aren't supporting the leader

The state machine should only campaign if it isn't supporting the
leader. The only exception is if the leader has explicitly asked a
follower to do so by initiating a campaignTransfer.

Epic: none

Release note: None

nvb committed:
raft: don't vote if we are supporting the leader

A follower should only vote for a candidate if it isn't supporting
a leader.

Epic: none
Release note: None

nvb committed:
kv,raft,storeliveness: fix a collection of bugs

nvb committed:
ui: add leader lease info to range debug page

nvb committed:
raft: disable faulty assertion

nvb committed:
sql: hack around #124987

nvb committed:
kv: add env var for leader leases

nvb committed:
kv: unquiesce on rejected lease acquisition

miraradeva committed:
storeliveness: add cluster setting to start and stop store liveness

The cluster setting `kv.storeliveness.enabled` starts and stops the main store liveness loop.

nvb committed:
raft: allow store liveness to be disabled

nvb committed:
kv: don't panic on nil raft status

nvb committed:
kv: use real HeartbeatInterval

nvb committed:
kv: patch store liveness shutdown

nvb committed:
storeliveness: add delay in transport
